### PR TITLE
Update SARIF output Rule Id to Opaque Rule Id

### PR DESCRIPTION
**Pull request by yongyan-gh**
As described in issue #1427

* Changed Rule Id value from "id" to "SPOTBUGS_HashCode(id)"
* Added name property to Rule, value is original rule "id" (BugPattern.type)
* Changed Result RuleId value to "SPOTBUGS_HashCode(ruleId)" accordingly

### CODE REVIEW

**Comment by yongyan-gh** [comment]
Attached sample sarif logs generated before/after the change for comparing. 
[sarif_result.zip](https://github.com/spotbugs/spotbugs/files/5974036/sarif_result.zip)

File: spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Result.java [hunk 1]
```diff
@@ -20,7 +20,7 @@
     final Level level;
 
     Result(@NonNull String ruleId, int ruleIndex, Message message, List<Location> locations, @NonNull Level level) {
-        this.ruleId = Objects.requireNonNull(ruleId);
+        this.ruleId = String.format("SPOTBUGS_%d", Objects.requireNonNull(ruleId).hashCode());
```
Comment (KengoTODA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Result.java:23]: Not sure the `rule id` can contain underscore or not 🤔 .
The [official doc](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#def_rule_id) says nothing about acceptable characters.

Comment (KengoTODA) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Result.java:23]: Note for test: hashcode could be negative.

**Review by KengoTODA — CHANGES_REQUESTED** [review]
Please consider removing the .vs directory. I'll test and confirm that generated id is valid and well-formatted.

**Comment by yongyan-gh** [comment]
> Please consider removing the .vs directory. I'll test and confirm that generated id is valid and well-formatted.

Removed, checked-in by mistake.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Result.java (hunk 1 above)
Comment (yongyan-gh) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Result.java:23]: The rule id can have underscore, e.g. "EC_UNRELATED_TYPES" is the rule id without the change in this PR.
The negative hash code is valid hash code, if change it e.g. math.abs may increase the possibility of a collision. 
cc @michaelcfanning @eddynaka what do you think about the opaque rule id?

Comment (michaelcfanning) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Result.java:23]: These ids need work. The prefix is too long. We shouldn't have underscores or hyphens in the name. Ideally, an opaque id starts with a well-known initial value, 1001 is typical, and then monotonically increases as each rule is added. So, something like this SB1001, SB1002, SB1003, would work best.

A useful convention is to retain rule ids of 0001 - 0999 for internal tool/engine errors.

I can appreciate that my idea requires a more intrusive change, we may need to consider creating a formal place for the rule id in the object model and to populate SARIF from this.

Comment (yongyan-gh) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Result.java:23]: In this case we probably need to assign the fixed Ids in meta data in SpotBugs and use these Ids as rule id, and the id always points to same rule.
E.g. in findbugs.xml
<BugPattern **ruleId="SB1001"** abbrev="SBSC" type="SBSC_USE_STRINGBUFFER_CONCATENATION" category="PERFORMANCE"/>
<BugPattern **ruleId="SB1002"**  abbrev="IIL" type="IIL_ELEMENTS_GET_LENGTH_IN_LOOP" category="PERFORMANCE" experimental="true"/>
...
Let me know what do you think?

**Comment by yongyan-gh** [comment]
close this PR since hash code id is not a good solution. Will work on the fix of another approach.

**Pull request closed without merging**